File: app/src/main/java/com/liyu/huahui/adapter/WordAdapter.java
package com.liyu.huahui.adapter;

import android.graphics.Paint;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.BaseViewHolder;
import com.liyu.huahui.R;
import com.liyu.huahui.model.Word;
import com.liyu.huahui.ui.MainActivity;
import com.liyu.huahui.utils.DownloadUtil;
import com.liyu.huahui.utils.Player;

import java.util.List;

/**
 * Created by liyu on 2017/3/2.
 */

public class WordAdapter extends BaseQuickAdapter<Word, BaseViewHolder> {

    private OnDataChangedListener listener;

    public WordAdapter(int layoutResId, List<Word> data) {
        super(layoutResId, data);
    }

    @Override
    protected void convert(final BaseViewHolder helper, final Word item) {
        helper.setText(R.id.tv_name, item.getName());

        helper.setText(R.id.tv_correct, "正确：" + item.getCorrectPhonetic());

        if (TextUtils.isEmpty(item.getWrongPhonetic())) {
            helper.setText(R.id.tv_wrong, item.getWrongPhonetic());
        } else {
            helper.setText(R.id.tv_wrong, "错误：" + item.getWrongPhonetic());
        }

        helper.setVisible(R.id.image_play, !TextUtils.isEmpty(item.getVoiceUrl()));

        ((TextView) helper.getView(R.id.tv_wrong)).getPaint()
                .setFlags(Paint.STRIKE_THRU_TEXT_FLAG | Paint.ANTI_ALIAS_FLAG);

        helper.getView(R.id.image_play).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                playVoice(item);
            }
        });

        helper.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                playVoice(item);
            }
        });

        helper.itemView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                final int deletedPosition = helper.getAdapterPosition();
                remove(deletedPosition);
                item.delete();
                if (listener != null) {
                    listener.onChanged(getData().size());
                }
                Snackbar.make(((MainActivity) mContext).getWindow().getDecorView().getRootView().findViewById(R.id.coordinatorLayout), "删除成功!",
                        Snackbar.LENGTH_LONG).setAction("撤销", new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        addData(deletedPosition, item);
                        item.save();
                        if (listener != null) {
                            listener.onChanged(getData().size());
                        }
                    }
                }).setActionTextColor(ContextCompat.getColor(mContext, R.color.snackbar_action_color)).show();
                return false;
            }
        });
    }

    private void playVoice(Word item) {
        DownloadUtil.start(item, new DownloadUtil.SingleFileDownloadListener() {
            @Override
            public void onCompleted(String path) {
                Player.getInstance().play(path);
            }

            @Override
            public void onFail(String msg) {
                Toast.makeText(mContext, msg, Toast.LENGTH_SHORT).show();
            }
        });

    }

    public interface OnDataChangedListener {
        void onChanged(int dateSize);
    }

    public void setDataChangedListener(OnDataChangedListener listener) {
        this.listener = listener;
    }
}

File: app/src/main/java/com/liyu/huahui/adapter/WordAdapter.kt
package com.liyu.huahui.adapter

import android.graphics.Paint
import android.text.TextUtils
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.viewholder.BaseViewHolder
import com.google.android.material.snackbar.Snackbar
import com.liyu.huahui.R
import com.liyu.huahui.model.Word
import com.liyu.huahui.ui.MainActivity
import com.liyu.huahui.utils.DownloadUtil.SingleFileDownloadListener
import com.liyu.huahui.utils.DownloadUtil.start
import com.liyu.huahui.utils.Player.Companion.instance

/**
 * Created by liyu on 2017/3/2.
 */
class WordAdapter(layoutResId: Int, data: List<Word>?) : BaseQuickAdapter<Word, BaseViewHolder>(layoutResId, data as MutableList<Word>?) {
    private var listener: OnDataChangedListener? = null
    override fun convert(helper: BaseViewHolder, item: Word) {
        helper.setText(R.id.tv_name, item.name)
        helper.setText(R.id.tv_correct, "英音：" + item.correctPhonetic)
        if (TextUtils.isEmpty(item.wrongPhonetic)) {
            helper.setText(R.id.tv_wrong, item.wrongPhonetic)
        } else {
            helper.setText(R.id.tv_wrong, "错误：" + item.wrongPhonetic)
        }
        helper.setVisible(R.id.image_play, !TextUtils.isEmpty(item.voiceUrl))
        (helper.getView<View>(R.id.tv_wrong) as TextView).paint.flags = Paint.STRIKE_THRU_TEXT_FLAG or Paint.ANTI_ALIAS_FLAG
        helper.getView<View>(R.id.image_play).setOnClickListener { playVoice(item) }
        helper.itemView.setOnClickListener { playVoice(item) }
        helper.itemView.setOnLongClickListener {
            val deletedPosition = helper.adapterPosition
            remove(deletedPosition)
            item.delete()
            if (listener != null) {
                listener!!.onChanged(data.size)
            }
            Snackbar.make((context as MainActivity).window.decorView.rootView.findViewById(R.id.coordinatorLayout), "删除成功!",
                    Snackbar.LENGTH_LONG).setAction("撤销") {
                addData(deletedPosition, item)
                item.save()
                if (listener != null) {
                    listener!!.onChanged(data.size)
                }
            }.setActionTextColor(ContextCompat.getColor(context, R.color.snackbar_action_color)).show()
            false
        }
    }

    private fun playVoice(item: Word) {
        start(context, item, object : SingleFileDownloadListener {
            override fun onCompleted(path: String?) {
                instance!!.play(path)
            }

            override fun onFail(msg: String?) {
                Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
            }
        })
    }

    interface OnDataChangedListener {
        fun onChanged(dateSize: Int)
    }

    fun setDataChangedListener(listener: OnDataChangedListener) {
        this.listener = listener
    }
}